src/internal/starlark/starlark.go

```diff
@@ -0,0 +1,232 @@
+// Package starlark runs Pachyderm-specific starlark programs.  It's recommended to import this as
+// "ourstar" when also using go.starlark.net/starlark in the same package.
+package starlark
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	starjson "go.starlark.net/lib/json"
+	starmath "go.starlark.net/lib/math"
+	startime "go.starlark.net/lib/time"
+	"go.starlark.net/starlark"
+	"go.starlark.net/syntax"
+	"go.uber.org/zap"
+)
+
+const (
+	goContextKey       = "goContext"
+	scriptDirectoryKey = "scriptDirectory"
+)
+
+func init() {
+	starlark.Universe["json"] = starjson.Module
+	starlark.Universe["time"] = startime.Module
+	starlark.Universe["math"] = starmath.Module
+}
+
+// GetContext returns a context.Context that applies to in the provided Starlark thread.
+func GetContext(t *starlark.Thread) context.Context {
+	x := t.Local(goContextKey)
+	if ctx, ok := x.(context.Context); ok {
+		return ctx
+	}
+	return pctx.TODO()
+}
+
+func scriptDirectory(t *starlark.Thread) string {
+	x := t.Local(scriptDirectoryKey)
+	if s, ok := x.(string); ok {
+		return s
+	}
+	return "/dev/null"
+}
+
+func nameScript(t *starlark.Thread, abs string) string {
+	wd, err := os.Getwd()
+	if err != nil {
+		return abs
+	}
+	rel, err := filepath.Rel(wd, abs)
+	if err != nil {
+		return abs
+	}
+	return rel
+}
+
+type loadResult struct {
+	globals starlark.StringDict
+	err     error
+}
+
+func newThread(name string) *starlark.Thread {
+	return &starlark.Thread{
+		Name: name,
+		Print: func(t *starlark.Thread, msg string) {
+			log.Info(GetContext(t), msg)
+		},
+	}
+}
+
+// Options controls the Starlark execution environment.
+type Options struct {
```

Comment:
If `Option` were a type that mutated an unexported `options` struct, then `Run` could have a signature like `(context.Context, string, RunFn, ...Option)`